### PR TITLE
feat(auth): add support for oauth2 with openid connect discovery

### LINKED ISSUE #4649: [auth][oauth] add initial oauth2 support

### LINKED ISSUE #4653: [oauth] support returnUrl query parameter

### LINKED ISSUE #4654: [oauth] add rate limiter to oauth endpoints

### LINKED ISSUE #4657: [auth][oauth] store state and nonce for oauth2 authorization code flow in cookies
Needed for multi-instance deployments of Phoenix.

### PR DESCRIPTION
**Pull request by axiomofjoy**
Adds initial support for OAuth2 identity providers that support the [OpenID Connect Discovery standard](https://openid.net/specs/openid-connect-discovery-1_0.html). From the standard:

`OpenID Providers supporting Discovery MUST make a JSON document available at the path formed by concatenating the string /.well-known/openid-configuration to the Issuer. The syntax and semantics of .well-known are defined in [RFC 5785](https://openid.net/specs/openid-connect-discovery-1_0.html#RFC5785) [RFC5785] and apply to the Issuer value when it contains no path component. openid-configuration MUST point to a JSON document compliant with this specification and MUST be returned using the application/json content type. The openid-configuration endpoint SHOULD support the use of [Cross-Origin Resource Sharing (CORS)](https://openid.net/specs/openid-connect-discovery-1_0.html#CORS) [CORS] and/or other methods as appropriate to enable JavaScript Clients and other Browser-Based Clients to access it.
`

This means that we can use environment variables to point Phoenix to the configuration at `/.well-known/openid-configuration` in order to know authorization server metadata such as the authorize and access token urls.

The environment variables currently required are:

- `PHOENIX_OAUTH2_<idp>_CLIENT_ID`
- `PHOENIX_OAUTH2_<idp>_CLIENT_SECRET`
- `PHOENIX_OAUTH2_<idp>_SERVER_METADATA_URL`

For the following IDPs, we provide logos on the login page:

- `GOOGLE`
- `MICROSOFT_ENTRA_ID` (previously known as Azure Active Directory)
- `AWS_COGNITO`

As a follow-up, we can add IDP-specific environment variables to help construct the server metadata url (e.g., using a tenant ID for Microsoft Entra ID).

resolves #4649
resolves #4653
resolves #4657
resolves #4654

### CODE REVIEW

File: src/phoenix/server/bearer_auth.py [hunk 1]
```diff
@@ -108,3 +127,35 @@ async def is_authenticated(request: Request) -> None:
         raise HTTPException(status_code=HTTP_401_UNAUTHORIZED, detail="Expired token")
     if claims.status is not ClaimSetStatus.VALID:
         raise HTTPException(status_code=HTTP_401_UNAUTHORIZED, detail="Invalid token")
+
+
+async def create_access_and_refresh_tokens(
```
Comment (RogerHYang) [inline src/phoenix/server/bearer_auth.py:132]: add a similar one for the oauth counterparts?

**Review by RogerHYang — APPROVED** [review]

File: src/phoenix/server/api/routers/oauth2.py [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+login_rate_limiter = ServerRateLimiter(
+    per_second_rate_limit=0.2,
+    enforcement_window_seconds=30,
+    partition_seconds=60,
+    active_partitions=2,
+)
+login_rate_limiter = fastapi_ip_rate_limiter(
+    login_rate_limiter,
+    paths=[
+        "oauth2/login",
```
[the remark this answers is not in the document]
Comment (axiomofjoy) [reply src/phoenix/server/api/routers/oauth2.py]: Changed to attach the rate limiter dependency on the route directly.

File: src/phoenix/db/migrations/versions/cd164e83824f_users_and_tokens.py [hunk 3]
```diff
@@ -69,6 +67,10 @@ def upgrade() -> None:
             nullable=True,
         ),
         sa.CheckConstraint("password_hash is null or password_salt is not null", name="salt"),
+        sa.UniqueConstraint(
+            "oauth2_client_id",
+            "oauth2_user_id",
```
[the remark this answers is not in the document]
Comment (axiomofjoy) [reply src/phoenix/db/migrations/versions/cd164e83824f_users_and_tokens.py:72]: That's a good idea.

File: src/phoenix/db/models.py [hunk 4]
```diff
@@ -637,12 +637,12 @@ class User(Base):
     role: Mapped["UserRole"] = relationship("UserRole", back_populates="users")
     username: Mapped[Optional[str]] = mapped_column(nullable=True, unique=True, index=True)
     email: Mapped[str] = mapped_column(nullable=False, unique=True, index=True)
-    auth_method: Mapped[str] = mapped_column(
```
[the remark this answers is not in the document]
Comment (axiomofjoy) [reply src/phoenix/db/models.py:640]: That's a good idea.

File: src/phoenix/server/api/routers/auth.py [hunk 5]
```diff
@@ -51,14 +52,14 @@
...
-        "/login",
-        "/logout",
-        "/refresh",
-        "/password-reset-email",
-        "/password-reset",
+        "/auth/login",
+        "/auth/logout",
+        "/auth/refresh",
+        "/auth/password-reset-email",
+        "/auth/password-reset",
```
[the remark this answers is not in the document]
Comment (axiomofjoy) [reply src/phoenix/server/api/routers/auth.py]: removed the explicit paths

File: src/phoenix/db/models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (axiomofjoy) [reply src/phoenix/db/models.py:640]: https://github.com/Arize-ai/phoenix/issues/4687

File: src/phoenix/db/migrations/versions/cd164e83824f_users_and_tokens.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (axiomofjoy) [reply src/phoenix/db/migrations/versions/cd164e83824f_users_and_tokens.py:72]: https://github.com/Arize-ai/phoenix/issues/4688

File: src/phoenix/server/bearer_auth.py (hunk 1 above)
Comment (axiomofjoy) [reply src/phoenix/server/bearer_auth.py:132]: Not sure what you have in mind here. This method is used inside the OAuth2 routers. Not sure we will re-use any of the code related to the OAuth2 authorization code flow elsewhere.

Comment (RogerHYang) [reply src/phoenix/server/bearer_auth.py:132]: ok. it was just a thought

**Pull request merged**